### PR TITLE
[MI-2222] Add pagination on subscriptions in rhs

### PR DESCRIPTION
**Pull request by ayusht2810**

### CODE REVIEW

File: webapp/src/containers/Rhs/index.tsx [hunk 1]
```diff
@@ -47,24 +52,32 @@ const Rhs = (): JSX.Element => {
         return {isLoading, isSuccess, isError, data: data as SubscriptionData[], error: (apiErr as FetchBaseQueryError)?.data as APIError | undefined};
     };
 
-    // Fetch all the subscriptions from the API
+    // Reset the page params and empty the subscription list
```
Comment (raghavaggarwal2308) [inline webapp/src/containers/Rhs/index.tsx:55]:
```suggestion
    // Reset the pagination params and empty the subscription list
```

File: webapp/src/containers/Rhs/index.tsx [hunk 2]
```diff
@@ -47,24 +52,32 @@ const Rhs = (): JSX.Element => {
...
         }
-        const subscriptionParams: FetchSubscriptionsParams = {page: Constants.DefaultPage, per_page: Constants.DefaultPageSize};
+
         setFetchSubscriptionParams(subscriptionParams);
         makeApiRequest(Constants.pluginApiServiceConfigs.fetchSubscriptions.apiServiceName, subscriptionParams);
-    }, [showAllSubscriptions]);
+    }, [paginationQueryParams]);
 
-    // Fetch subscriptions from the API if the active channel changes or showAllSubscription is reset
+    // Reset states on changing channel or using toggle
```
Comment (raghavaggarwal2308) [inline webapp/src/containers/Rhs/index.tsx:78]:
```suggestion
    // Reset states on changing channel or using toggle switch
```

File: webapp/src/containers/Rhs/rhsData.tsx [hunk 3]
```diff
@@ -67,6 +74,10 @@ const RhsData = ({
         list: subscription.subscription_events.split(',').map((event) => SubscriptionEventLabels[event as SubscriptionEvents]),
     }), []);
 
+    const hasMoreSubscriptions = useMemo<boolean>(() => (
+        subscriptionList.length !== 0 && (subscriptionList.length - (paginationQueryParams.page * Constants.DefaultPageSize) === Constants.DefaultPageSize)
+    ), [subscriptionList]);
```
Comment (raghavaggarwal2308) [inline webapp/src/containers/Rhs/rhsData.tsx:79]: shouldn't this contain paginationQuersParams as well?

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

**Review by raghavaggarwal2308 — APPROVED** [review]

File: webapp/src/components/spinner/index.tsx [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+
+import './styles.scss';
+
+const Spinner = ({extraClass = ''}: { extraClass?: string }): JSX.Element => (
```
Comment (muazzamqazi) [inline webapp/src/components/spinner/index.tsx:5]:
```suggestion
interface SpinnerProps { extraClass?: string };

const Spinner = ({extraClass = ''}: SpinnerProps): JSX.Element => (
```

File: webapp/src/containers/Rhs/index.tsx [hunk 5]
```diff
@@ -47,24 +52,32 @@ const Rhs = (): JSX.Element => {
         return {isLoading, isSuccess, isError, data: data as SubscriptionData[], error: (apiErr as FetchBaseQueryError)?.data as APIError | undefined};
     };
 
-    // Fetch all the subscriptions from the API
+    // Reset the pagination params and empty the subscription list
+    const resetStates = useCallback(() => {
+        setPaginationQueryParams(() => ({page: Constants.DefaultPage, per_page: Constants.DefaultPageSize}));
+        setSubscriptionList(() => []);
```
Comment (muazzamqazi) [inline webapp/src/containers/Rhs/index.tsx:58]:
```suggestion
        setSubscriptionList([]);
```

File: webapp/src/containers/Rhs/rhsData.tsx [hunk 6]
```diff
@@ -104,41 +115,57 @@ const RhsData = ({
...
+                className='rhs-content__cards-container'
+            >
+                {subscriptionList.length > 0 && (
+                    <InfiniteScroll
+                        dataLength={subscriptionList.length}
+                        next={handlePagination}
+                        hasMore={hasMoreSubscriptions}
+                        loader={<Spinner/>}
+                        endMessage={
+                            <p style={{textAlign: 'center'}}>
```
Comment (muazzamqazi) [inline webapp/src/containers/Rhs/rhsData.tsx:129]: Can we use Bootstrap classes and avoid in-line styles?

**Review by muazzamqazi — APPROVED** [review]

File: webapp/src/components/spinner/index.tsx [hunk 7]
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+
+import './styles.scss';
+
+interface SpinnerProps { extraClass?: string }
```
Comment (manojmalik20) [inline webapp/src/components/spinner/index.tsx:5]: Use proper indentation here.

File: webapp/src/components/spinner/index.tsx [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+const Spinner = ({extraClass = ''}: SpinnerProps): JSX.Element => (
```
Comment (manojmalik20) [inline webapp/src/components/spinner/index.tsx:7]: Shouldn't this component be in the UI library?

File: webapp/src/containers/Rhs/index.tsx [hunk 9]
```diff
@@ -47,24 +53,38 @@ const Rhs = (): JSX.Element => {
         return {isLoading, isSuccess, isError, data: data as SubscriptionData[], error: (apiErr as FetchBaseQueryError)?.data as APIError | undefined};
     };
 
-    // Fetch all the subscriptions from the API
+    // Reset the pagination params and empty the subscription list
+    const resetStates = useCallback(() => {
+        setPaginationQueryParams(() => ({page: Constants.DefaultPage, per_page: Constants.DefaultPageSize}));
```
Comment (manojmalik20) [inline webapp/src/containers/Rhs/index.tsx:58]:
```suggestion
        setPaginationQueryParams({page: Constants.DefaultPage, per_page: Constants.DefaultPageSize});
```
What's the need of providing a function here?

File: webapp/src/containers/Rhs/index.tsx [hunk 10]
```diff
@@ -174,24 +188,30 @@ const Rhs = (): JSX.Element => {
                 setSubscriptionsAuthorized(true);
             }
         }
-    }, [getSubscriptionsState().isError, getSubscriptionsState().isSuccess]);
+
+        if (!subscriptionsState.isLoading && subscriptionsState.isSuccess) {
```
Comment (manojmalik20) [inline webapp/src/containers/Rhs/index.tsx:192]: Move this `if` block inside the above `if` block.

File: webapp/src/containers/Rhs/rhsData.tsx [hunk 11]
```diff
@@ -67,6 +74,10 @@ const RhsData = ({
         list: subscription.subscription_events.split(',').map((event) => SubscriptionEventLabels[event as SubscriptionEvents]),
     }), []);
 
+    const hasMoreSubscriptions = useMemo<boolean>(() => (
+        subscriptionList.length !== 0 && (subscriptionList.length - (paginationQueryParams.page * Constants.DefaultPageSize) === Constants.DefaultPageSize)
```
Comment (manojmalik20) [inline webapp/src/containers/Rhs/rhsData.tsx:78]:
```suggestion
        subscriptionList.length > 0 && (subscriptionList.length - (paginationQueryParams.page * Constants.DefaultPageSize) === Constants.DefaultPageSize)
```
Can you explain this logic? I think it's wrong. :thinking:

File: webapp/src/components/spinner/index.tsx (hunk 8 above)
Comment (ayusht2810) [reply webapp/src/components/spinner/index.tsx:9]: the spinner used in mm is different from ours so for now we kept it in SN

File: webapp/src/containers/Rhs/rhsData.tsx (hunk 11 above)
Comment (ayusht2810) [reply webapp/src/containers/Rhs/rhsData.tsx]: updated the logic now

File: webapp/src/containers/Rhs/index.tsx [hunk 12]
```diff
@@ -176,6 +176,7 @@ const Rhs = (): JSX.Element => {
         }
 
         if (subscriptionsState.isSuccess) {
+            setSubscriptionList([...subscriptionList, ...subscriptions]);
```
Comment (manojmalik20) [inline webapp/src/containers/Rhs/index.tsx:179]: Don't you need to check for `!isLoading`?

File: webapp/src/containers/Rhs/rhsData.tsx [hunk 13]
```diff
@@ -75,7 +75,7 @@ const RhsData = ({
     }), []);
 
     const hasMoreSubscriptions = useMemo<boolean>(() => (
-        subscriptionList.length !== 0 && (subscriptionList.length - (paginationQueryParams.page * Constants.DefaultPageSize) === Constants.DefaultPageSize)
+        (subscriptionList.length - (paginationQueryParams.page * Constants.DefaultPageSize) >= Constants.DefaultPageSize)
```
Comment (manojmalik20) [inline webapp/src/containers/Rhs/rhsData.tsx:78]: I think it's still wrong. Correct would be - 
```suggestion
        subscriptionList.length === Constants.DefaultPageSize
```

Comment (ayusht2810) [reply webapp/src/containers/Rhs/rhsData.tsx:78]: on thinking more I guess this logic was already correct `(subscriptionList.length - (paginationQueryParams.page * Constants.DefaultPageSize) === Constants.DefaultPageSize)` and `subscriptionList` is the total length of subscriptions, not just the list of current subscriptions fetched

File: webapp/src/containers/Rhs/index.tsx (hunk 12 above)
Comment (manojmalik20) [reply webapp/src/containers/Rhs/index.tsx:179]: Okay, but you should still test this thoroughly.

File: webapp/src/containers/Rhs/rhsData.tsx (hunk 13 above)
Comment (manojmalik20) [reply webapp/src/containers/Rhs/rhsData.tsx:78]: I think you should change the variable name to `totalSubscriptions`.

**Review by manojmalik20 — APPROVED** [review]

**Pull request merged**